tiledb/sm/query/sparse_unordered_with_dups_reader.cc

```diff
@@ -564,320 +397,700 @@ Status SparseUnorderedWithDupsReader::create_result_tiles() {
+
+      max_num_cells = std::min(max_num_cells, temp_num_cells);
+    } else {
+      auto temp_num_cells = size / array_schema_->cell_size(name);
+      max_num_cells = std::min(max_num_cells, temp_num_cells);
+    }
+  }
+
+  // User gave us some empty buffers, exit.
+  if (max_num_cells == 0) {
+    zero_out_buffer_sizes();
+    return Status::Ok();
+  }
+
+  // Compute initial bound for result tiles.
+  uint64_t rt_idx = 0;
+  uint64_t cell_offset = 0;
+  for (auto& rt : result_tiles_[0]) {
+    const auto cell_num =
+        subarray_.is_set() ? rt.bitmap_num_cells : rt.cell_num();
+    if (cell_offset + cell_num > max_num_cells)
+      break;
+
+    rt_idx++;
+  }
+
+  // Calculate memory budget.
+  uint64_t memory_budget_copy = memory_budget_ - memory_used_qc_tiles_total_ -
+                                memory_used_result_tile_ranges_ -
+                                array_memory_tracker_->get_memory_usage();
+
+  // Make sure we respect memory budget for copy operation.
+  std::mutex it_mtx;
+  auto buffers_it = buffers_.begin();
+  auto status = parallel_for(
+      storage_manager_->compute_tp(), 0, buffers_.size(), [&](uint64_t) {
+        std::unordered_map<std::string, tiledb::sm::QueryBuffer>::iterator it;
+        {
+          std::unique_lock<std::mutex> ul(it_mtx);
+          it = buffers_it++;
         }
-        tile_used = true;
+        const auto& name = it->first;
+        const auto var_sized = array_schema_->var_size(name);
+        if (subarray_.is_set() && array_schema_->is_dim(name))
+          return Status::Ok();
+
+        uint64_t mem_usage = 0;
+        auto rt = result_tiles_[0].begin();
+        uint64_t idx = 0;
+        while (rt != result_tiles_[0].end()) {
+          uint64_t tile_size = 0;
+          RETURN_NOT_OK(get_attribute_tile_size(
+              name, rt->frag_idx(), rt->tile_idx(), &tile_size));
+
+          // Account for the pointers to the var data.
+          if (var_sized) {
+            auto cell_num =
+                subarray_.is_set() ? rt->bitmap_num_cells : rt->cell_num();
+            tile_size += sizeof(void*) * cell_num;
+          }
+
+          if (mem_usage + tile_size > memory_budget_copy) {
+            break;
+          }
+          mem_usage += tile_size;
+          rt++;
+          idx++;
+        }
+
+        {
+          std::unique_lock<std::mutex> ul(it_mtx);
+          rt_idx = std::min(idx, rt_idx);
+        }
+
+        return Status::Ok();
+      });
+  RETURN_NOT_OK_ELSE(status, logger_->status(status));
+
+  if (rt_idx == 0)
+    return Status::SparseUnorderedWithDupsReaderError(
+        "Unable to copy one tile with current budget/buffers");
+
+  *max_rt_idx = rt_idx;
+
+  return Status::Ok();
+}
+
+template <class BitmapType>
+template <class OffType>
+Status SparseUnorderedWithDupsReader<BitmapType>::copy_tiles() {
+  auto timer_se = stats_->start_timer("copy_tiles");
+
+  uint64_t max_rt_idx = 0;
+  RETURN_NOT_OK(initial_copy_bound(&max_rt_idx));
```

Comment:
Some documentation needed here, e.g., we are calculating the maximum number of tiles we will copy from to the user buffers based on the memory budget.